src/BaseListView.php

```diff
@@ -1,158 +1,195 @@
-     *            The "tag" element specifies the tag name of the emptyText element and defaults to "div".
-     *
-     * @see \Yiisoft\Html\Html::renderTagAttributes() for details on how attributes are being rendered.
+     * The "tag" element specifies the tag name of the emptyText element and defaults to "div".
+     * @see Html::renderTagAttributes() for details on how attributes are being rendered.
      */
-    public $emptyTextOptions = ['class' => 'empty'];
+    protected array $emptyTextOptions = ['class' => 'empty'];
     /**
      * @var string the layout that determines how different sections of the list view should be organized.
-     *             The following tokens will be replaced with the corresponding section contents:
-     *
-     * - `{summary}`: the summary section. See [[renderSummary()]].
-     * - `{items}`: the list items. See [[renderItems()]].
-     * - `{sorter}`: the sorter. See [[renderSorter()]].
-     * - `{pager}`: the pager. See [[renderPager()]].
+     * The following tokens will be replaced with the corresponding section contents:
+     * - `{summary}`: the summary section. See {@see renderSummary()}.
+     * - `{items}`: the list items. See {@see renderItems()}.
+     * - `{sorter}`: the sorter. See {@see renderSorter()}.
+     * - `{pager}`: the pager. See {@see renderPager()}.
      */
-    public $layout = "{summary}\n{items}\n{pager}";
+    protected string $layout = "{summary}\n{items}\n{pager}";
     /**
-     * @var \Yiisoft\I18n\MessageFormatterInterface
+     * @var MessageFormatterInterface
      */
-    private $formatter;
+    private static MessageFormatterInterface $messageFormatter;
+    /**
+     * @var View
+     */
+    private static View $view;
+    /**
+     * @var Aliases
+     */
+    private static Aliases $aliases;
+
+    public function __construct(MessageFormatterInterface $messageFormatter, View $view, Aliases $aliases)
+    {
+        self::$messageFormatter = $messageFormatter;
+        self::$view = $view;
+        self::$aliases = $aliases;
+    }
 
     /**
      * Renders the data models.
      *
      * @return string the rendering result.
      */
-    abstract public function renderItems();
+    abstract public function renderItems(): string;
+
+    /**
+     * @return CountableDataInterface|DataReaderInterface|FilterableDataInterface|OffsetableDataInterface|SortableDataInterface
+     */
+    public function getDataReader()
+    {
+        return $this->dataReader;
+    }
+
+    protected function getAliases(): Aliases
+    {
+        return self::$aliases;
+    }
 
-    public function __construct(MessageFormatterInterface $formatter)
+    public function showEmptyText(bool $value): self
     {
-        $this->formatter = $formatter;
+        $this->showEmptyText = $value;
+
+        return $this;
     }
 
     /**
      * Initializes the view.
+     *
+     * @throws InvalidConfigException
      */
-    public function init(): void
+    public function init(): self
```

Comment:
Can it be done in constructor?

Reply:
I don't think so. I was try and one test is failed